### PR TITLE
New configuration for abilities to raise the priority of mutations.

### PR DESCRIPTION
**Pull request by maskshell**
In some scenarios, a lots of mutations in an busy cluster will affect
the server normal operation due to blocking the queue.
We may want increase the process speed of mutations.

But in current design, mutation seems has the lowest priority.

So we add a new configuration item to separate mutation and merge.

* for merge. `max_bytes_to_merge_at_max_space_in_pool`
* for mutations. `max_bytes_to_mutate_at_max_space_in_pool`

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):
New configuration for abilities to raise the priority of mutations.

In some scenarios, a lots of mutations in an busy cluster will affect
the server normal operation due to blocking the queue.
We may want increase the process speed of mutations.

But in current design, mutation seems has the lowest priority.

So we add a new configuration item to separate mutation and merge.

* for merge. `max_bytes_to_merge_at_max_space_in_pool`
* for mutations. `max_bytes_to_mutate_at_max_space_in_pool`
...

Detailed description (optional):

...

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Could you please take a look to `ClickHouse build check — 0/14 builds are OK`

**Comment by sunadm** [comment]
> Could you please take a look to `ClickHouse build check — 0/14 builds are OK`?

My fault. Already fixed.

Could u pls tell me how to fix the failed of "Style check" ?

File: dbms/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp [hunk 1]
```diff
@@ -974,9 +974,19 @@ bool ReplicatedMergeTreeQueue::shouldExecuteLogEntry(
           * because the leader replica does not assign merges of greater size (except OPTIMIZE PARTITION and OPTIMIZE FINAL).
           */
         const auto data_settings = data.getSettings();
-        bool ignore_max_size = (entry.type == LogEntry::MERGE_PARTS) && (max_source_parts_size == data_settings->max_bytes_to_merge_at_max_space_in_pool);
+        bool ignore_max_size_for_merge  = (entry.type == LogEntry::MERGE_PARTS) && (max_source_parts_size == data_settings->max_bytes_to_merge_at_max_space_in_pool);
+        bool ignore_max_size_for_mutate = (entry.type == LogEntry::MUTATE_PART) && (max_source_parts_size == data_settings->max_bytes_to_mutate_at_max_space_in_pool);
```
Comment (tavplubix) [inline dbms/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp]: Please take a look at `MergeTreeDataMergerMutator::getMaxSourcePartSizeForMutation()`. It calculates `max_source_parts_size` as `disk_space / DISK_USAGE_COEFFICIENT_TO_RESERVE`, so it almost never equals to `max_bytes_to_mutate_at_max_space_in_pool` and `ignore_max_size_for_mutate` is false

**Review by tavplubix — CHANGES_REQUESTED** [review]

**Comment by alesapin** [comment]
Actually your idea (give mutations higher priority) is already implemented. First of all, you can tune `max_replicated_mutations_in_queue` and `max_replicated_merges_in_queue`. It allows you to have as many mutations in the queue as you want. Another setting is `number_of_free_entries_in_pool_to_execute_mutation`. It has quite a big value by default, but if you choose the smaller one, your mutations will take up more threads in background pool. And if you want to execute them as fast as possible you should also enlarge `background_pool_size setting`.

You can see an example of tuning of these settings in my test https://github.com/clickhouse/ClickHouse/blob/master/tests/queries/0_stateless/01079_parallel_alter_modify_zookeeper.sh#L13. 

Your implementation in this PR is not correct, because part size for mutations is unlimited (the only limit is available disk space). Such a setting doesn't make sense. 

If I understood your idea incorrectly fill free to reopen this pull request. Thank you, for your contribution, new ideas are welcome.

**Comment by silviucpp** [comment]
@alesapin I think https://github.com/ClickHouse/ClickHouse/issues/10897 is related to this problem as well. In case number_of_free_entries_in_pool_to_execute_mutationis > than background_pool_size the mutations will block forever. So at least some alerts should be added in logs.

**Pull request closed without merging**